### PR TITLE
YoloV5 inference support of ONNX models

### PR DESCRIPTION
**Pull request by thhart**
## Description ##
adds support for YoloV5 object detection models via ONNX

### Changes ###
added YoloV5Translator with following features: input squeezer, output parser and non max suppression boxes

### Howto ###
YoloV5 supports at least two output formats, to use models here it is important to specify model.model[-1].export = False  in export.py.
An ONNX export can be done like this: https://github.com/ultralytics/yolov5/issues/251

## Example ##

```
         final YoloV5Translator translator = new YoloV5Translator(YoloV5Translator.builder().optSynsetArtifactName("synset.txt"));
         Criteria<Image, DetectedObjects> criteria = Criteria.builder()
               .setTypes(Image.class, DetectedObjects.class) // defines input and output data type
               .optDevice(Device.cpu())
               .optTranslator(translator)
               .optModelUrls("file:///path-to-model/") // search models in specified path
               .optModelName("yolov5-exported-model.onnx")
               .optEngine("OnnxRuntime")
               .build();
         final ZooModel<Image, DetectedObjects> model = ModelZoo.loadModel(criteria);
         Predictor<Image, DetectedObjects> predictor = model.newPredictor(translator);
         final File input = new File("a-file-to-analyze.jpg");
         BufferedImageFactory factory = new BufferedImageFactory();
         final BufferedImage ix = ImageIO.read(input); //should be scaled or padded to model dims
         DetectedObjects detection = predictor.predict(factory.fromImage(ix));
```

### CODE REVIEW

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+      return nmsThresh;
+   }
+
+   public void setNmsThresh(double nmsThresh) {
+      this.nmsThresh = nmsThresh;
+   }
+
+   protected DetectedObjects nms(ArrayList<IntermediateResult> list) {
+      List<String> retClasses = new ArrayList<>();
+      List<java.lang.Double> retProbs = new ArrayList<>();
```
Comment (stu1130) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:100]: nit: 
```suggestion
      List<Double> retProbs = new ArrayList<>();
```

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (stu1130) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:2]: nit:2020

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+      List<BoundingBox> retBB = new ArrayList<>();
+
+      for (int k = 0; k < classes.size(); k++) {
+         //1.find max confidence per class
+         PriorityQueue<IntermediateResult> pq =
+               new PriorityQueue<>(
+                     50,
+                     (lhs, rhs) -> {
+                        // Intentionally reversed to put high confidence at the head of the queue.
+                        return java.lang.Double.compare(rhs.getConfidence(), lhs.getConfidence());
```
Comment (stu1130) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:110]:
```suggestion
                        Double.compare(rhs.getConfidence(), lhs.getConfidence());
```

**Review by stu1130 — COMMENTED** [review]
Thanks for your contribution!
Could you run
`./gradlew fJ build`

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.modality.cv.translator;
+
+import java.util.*;
+import java.awt.geom.Rectangle2D;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:16]: We should avoid using awt package if possible, awt package is not portable. It doesn't work on Android.
You can use: ai.djl.modality.cv.output.Rectangle

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+import java.awt.geom.Rectangle2D.Double;
+import ai.djl.modality.cv.output.*;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:18]: Avoid wildcard import

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+/**
+ A translator for yolo V5 models. This was tested with ONNX exported Yolo models.
+ For details check here: https://github.com/ultralytics/yolov5
+ */
+public class YoloV5Translator extends ObjectDetectionTranslator {
+   private final YOLO_OUTPUT_TYPE yoloOutputLayerType;
+   private double nmsThresh = 0.4F;
+
+   public enum YOLO_OUTPUT_TYPE {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:31]: 1. enum name should be: YolyOuputType
2. class should be at the end of the class,

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+import ai.djl.ndarray.types.DataType;
+import ai.djl.translate.*;
+
+/**
+ A translator for yolo V5 models. This was tested with ONNX exported Yolo models.
+ For details check here: https://github.com/ultralytics/yolov5
+ */
+public class YoloV5Translator extends ObjectDetectionTranslator {
+   private final YOLO_OUTPUT_TYPE yoloOutputLayerType;
+   private double nmsThresh = 0.4F;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:29]:
```suggestion
   private double nmsThresh = 0.4;
```

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+import ai.djl.ndarray.*;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.translate.*;
+
+/**
+ A translator for yolo V5 models. This was tested with ONNX exported Yolo models.
+ For details check here: https://github.com/ultralytics/yolov5
+ */
+public class YoloV5Translator extends ObjectDetectionTranslator {
+   private final YOLO_OUTPUT_TYPE yoloOutputLayerType;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:28]: avoid final member modifier for simplicity and readability.

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+import java.util.*;
+import java.awt.geom.Rectangle2D;
+import java.awt.geom.Rectangle2D.Double;
+import ai.djl.modality.cv.output.*;
+import ai.djl.ndarray.*;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.translate.*;
+
+/**
+ A translator for yolo V5 models. This was tested with ONNX exported Yolo models.
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:24]:
```suggestion
 * A translator for yolo V5 models. This was tested with ONNX exported Yolo models.
```

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+ For details check here: https://github.com/ultralytics/yolov5
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:25]: Java doc requires * at each line

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+   protected double boxIou(Rectangle2D a, Rectangle2D b) {
+      return boxIntersection(a, b) / boxUnion(a, b);
+   }
+
+   protected double boxUnion(Rectangle2D a, Rectangle2D b) {
+      double i = boxIntersection(a, b);
+      return (a.getMaxX() - a.getMinX()) * (a.getMaxY() - a.getMinY()) + (b.getMaxX() - b.getMinX()) * (b.getMaxY() - b.getMinY()) - i;
+   }
+
+   public double getNmsThresh() {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:90]: public method requires JavaDoc

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+      return nmsThresh;
+   }
+
+   public void setNmsThresh(double nmsThresh) {
+      this.nmsThresh = nmsThresh;
+   }
+
+   protected DetectedObjects nms(ArrayList<IntermediateResult> list) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:98]:
```suggestion
   protected DetectedObjects nms(List<IntermediateResult> list) {
```

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+      double l2 = x2 - w2 / 2;
+      double left = Math.max(l1, l2);
+      double r1 = x1 + w1 / 2;
+      double r2 = x2 + w2 / 2;
+      double right = Math.min(r1, r2);
+      return right - left;
+   }
+
+   private DetectedObjects processFromBoxOutput(TranslatorContext ctx, NDList list) {
+      final NDArray ndArray = list.get(0);
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:152]: Avoid "final" for local variable

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+      return processFromBoxOutput(ctx, list);
+   }
+
+   /**
+    The builder for {@link YoloV5Translator}.
+    */
+   public static class Builder extends ObjectDetectionBuilder<YoloV5Translator.Builder> {
+      public Builder() {
+         // custom pipeline to match default YoloV5 input layer
+         pipeline = new Pipeline().add(array -> array.transpose(2, 0, 1).toType(DataType.FLOAT32, false).div(255));
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:206]: Move this to build() method.
pipeline could be manually set in the builder, We should create pipeline only if user didn't set pipeline.

**Review by frankfliu — CHANGES_REQUESTED** [review]

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 15]
```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+        private int detectedClass;
+        /**
+         * A unique identifier for what has been recognized. Specific to the class, not the instance
+         * of the object.
+         */
+        private String id;
+        /** Optional location within the source image for the location of the recognized object. */
+        private Rectangle location;
+
+        private IntermediateResult(
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/cv/translator/YoloV5Translator.java:283]:
```suggestion
       IntermediateResult(
```

**Review by frankfliu — COMMENTED** [review]
This looks good to me.

Would you mind import this mode into ONNX runtime model zoo in separate PR?

**Comment by thhart** [comment]
> Would you mind import this mode into ONNX runtime model zoo in separate PR?
Please note I use DJL with own models however I can understand it can be good to have a sample model available and also for testing maybe, if I find some time I will take a standard model.

**Review by frankfliu — APPROVED** [review]

**Comment by hongyaohongyao** [comment]
cost memory leak

**Pull request merged**